specs/agents/tracing-instrumentation-aws-lambda.md

```diff
@@ -200,3 +200,7 @@ Field | Value | Description | Source
 Lambda functions are immediately frozen as soon as the handler method ends. In case APM data is sent in an asyncronous way (as most of the agents do by default) data can get lost if not sent before the lambda function ends.
 
 Therefore, the Lambda instrumentation has to ensure that data is flushed in a blocking way before the execution of the handler function ends.
+
+# Flushed Signal
+Some Lambda functions will use the custom-built Lambda extension that allows the agent to send its data locally. The extension asynchronously forwards the data it receives from the agent to the APM server so the Lambda function can return its result with minimal delay. In order for the extension to know when it can flush its data, it must receive a signal indicating that the lambda function has completed. There are two possible signals: one is via a subscription to the AWS Lambda Logs API and the other is an agent intake request with the query param `flushed=true`. A signal from the agent is preferrable because there is an inherent delay with the sending of the Logs API signal.
+Therefore, the agent should send its final intake request at the end of the function invocation with the query param `flushed=true`. In case there is no more data to send at the end of the function invocation, the agent should send an empty intake request with this query param.
```

Comment:
I'd even go as far as spec'ing this as a MUST, then we can deprecate and phase out flushing based on the Logs API signal.

Reply:
> I'd even go as far as spec'ing this as a MUST, then we can deprecate and phase out flushing based on the Logs API signal.

Can we assume that an APM agent can always detect the end of lambda function?

Reply:
I haven't instrumented lambda functions in the Ruby agent but in this spec under `Generic Lambda Instrumentation`, it says:

> In general, to instrument Lambda functions, we create transactions that wrap the execution of that handler method

so the agent should know that the function is complete in the same way that it knows to end the transaction.

Reply:
Maybe I don't not fully understand what is meant by "end of lambda function" - is it an instant before lambda infrastructure terminates/suspends the spawned application process?

Reply:
Yes, as this spec is about dedicated auto-instrumentation support for Lambda.

Reply:
> I haven't instrumented lambda functions in the Ruby agent but in this spec under `Generic Lambda Instrumentation`, it says:
> 
> > In general, to instrument Lambda functions, we create transactions that wrap the execution of that handler method
> 
> so the agent should know that the function is complete in the same way that it knows to end the transaction.

Please let me see if I understood it correctly: so agent should add `flushed=true` when the spawned application process finished handling a client request and the process might be terminated/suspended by lambda infrastructure even though lambda infrastructure might not actually terminate/suspend the process and instead direct even more client requests for it handle. Did I understand it correctly?

Reply:
That's correct. That's because the agent has no way of knowing whether the Lambda will freeze after the invocation or if another request is about to be handled by the same Lambda instance.